### PR TITLE
Update: responses and code status

### PR DESCRIPTION
**Pull request by Amna-Alshawa**
## Feature description
- Changes in the code status and adding success section in the API response for Shift permission API.
- Add missing part in the checkin-deadline method.
- notification list api.

## Output screenshots (optional)
1- Created Successfully:
Screen Shot 2021-11-08 at 1 23 01 PM

2- Duplicated  record:
Screen Shot 2021-11-08 at 1 20 59 PM

3- Not created successfully:
Screen Shot 2021-11-08 at 1 23 19 PM
4- 500 status code response:
Screen Shot 2021-11-08 at 4 28 29 PM

5- Notification list:
Screen Shot 2021-11-08 at 4 23 40 PM

## Is there any existing behavior change of other features due to this code change?
No

## Was this feature tested on all the browsers?
  - [x] Chrome
  - [ ] Safari

### CODE REVIEW

File: one_fm/api/mobile/shift_permission.py [hunk 1]
```diff
@@ -31,17 +31,17 @@ def create_shift_permission(employee, permission_type, date, reason, leaving_tim
...
+                return response("{0} has already applied for permission to {1} on {2}.".format(doc.emp_name, permission_type.lower(), date), {}, False, 409)
 
         elif not shift or not type or not assigned_shift or not shift_supervisor:
-            return response({'message':"Shift Details are missing, Please Make sure You are have a Scheduled record on {0}".format(date),'data':{}},400)
+            return response("Shift Permission Not Created Successfully", {}, False, 400)
            
     except Exception as e:
         frappe.log_error(frappe.get_traceback())
-        return response({'message':"Shift Permission Not Created Successfully",'data':{}},500)
+        return response("Shift Permission Not Created Successfully", {}, False, 500)
```
Comment (Ehtesham599) [inline one_fm/api/mobile/shift_permission.py]: Set response message as the Exception message

File: one_fm/api/mobile/shift_permission.py [hunk 2]
```diff
@@ -31,17 +31,17 @@ def create_shift_permission(employee, permission_type, date, reason, leaving_tim
...
-                return response({'message':"Shift Permission Successfully Created",'data':doc},201)
+                return response("Shift Permission Successfully Created",doc,True,201)
             elif has_duplicate:
                 doc = frappe.get_doc('Shift Permission',{"employee": employee, "date":date, "assigned_shift": assigned_shift, "permission_type": permission_type})
-                return response({'message':"{0} has already applied for permission to {1} on {2}.".format(doc.emp_name, permission_type.lower(), date),'data':{}},409)
+                return response("{0} has already applied for permission to {1} on {2}.".format(doc.emp_name, permission_type.lower(), date), {}, False, 409)
 
         elif not shift or not type or not assigned_shift or not shift_supervisor:
-            return response({'message':"Shift Details are missing, Please Make sure You are have a Scheduled record on {0}".format(date),'data':{}},400)
+            return response("Shift Permission Not Created Successfully", {}, False, 400)
```
Comment (Ehtesham599) [inline one_fm/api/mobile/shift_permission.py]: Maybe change the wording? Something like "Could not create Shift Permission".

Comment (Ehtesham599) [reply one_fm/api/mobile/shift_permission.py]: Also, provide appropriate reason as to why Shift permission was not able to be created.

File: one_fm/api/mobile/shift_permission.py [hunk 3]
```diff
@@ -116,14 +116,14 @@ def shift_permission_approved(supervisor_id, shift_permission_id):
...
             elif shift_permission_record.workflow_state == "Approved":
-                return response({'message':"Shift Permission is Already Approved",'data':{}},400)
+                return response("Shift Permission is Already Approved", {}, False, 400)
         elif shift_supervisor != supervisor_id:
-            return response({'message':"Only Supervisor Has The Right to Approve",'data':{}},400)
+            return response("Only Supervisor Has The Right to Approve", {}, False, 400)
     except Exception as e:
         frappe.log_error(frappe.get_traceback())
-        return response({'message':"Shift Permission Not Approved Successfully",'data':{}},500)
+        return response("Shift Permission Not Approved Successfully", {}, False, 500)
```
Comment (Ehtesham599) [inline one_fm/api/mobile/shift_permission.py]: Return Exception message as response.

File: one_fm/api/mobile/shift_permission.py [hunk 4]
```diff
@@ -142,14 +142,14 @@ def shift_permission_rejected(supervisor_id, shift_permission_id):
...
             elif shift_permission_record.workflow_state == "Rejected":
-                return response({'message':"Shift Permission is Already Rejected",'data':{}},400)
+                return response("Shift Permission is Already Rejected", {}, False, 400)
         elif shift_supervisor != supervisor_id:
-            return response({'message':"Only Supervisor Has The Right to Reject",'data':{}},400)
+            return response("Only Supervisor Has The Right to Reject", {}, False, 400)
     except Exception as e:
         frappe.log_error(frappe.get_traceback())
-        return response({'message':"Shift Permission Not Rejected Successfully",'data':{}},500)
+        return response("Shift Permission Not Rejected Successfully", {}, False, 500)
```
Comment (Ehtesham599) [inline one_fm/api/mobile/shift_permission.py]: Return exception message as response.

File: one_fm/api/mobile/shift_permission.py [hunk 5]
```diff
@@ -169,14 +169,14 @@ def shift_permission_cancelled(supervisor_id, shift_permission_id):
...
             elif shift_permission_record.workflow_state == "Cancelled":
-                return response({'message':"Shift Permission is Already Cancelled",'data':{}},400)
+                return response("Shift Permission is Already Cancelled", {}, False, 400)
         elif shift_supervisor != supervisor_id:
-            return response({'message':"Only Supervisor Has The Right to Cancel",'data':{}},400)
+            return response("Only Supervisor Has The Right to Cancel", {}, False, 400)
     except Exception as e:
         frappe.log_error(frappe.get_traceback())
-        return response({'message':"Shift Permission Not Cancelled Successfully",'data':{}},500)
+        return response("Shift Permission Not Cancelled Successfully", {}, False, 500)
```
Comment (Ehtesham599) [inline one_fm/api/mobile/shift_permission.py]: Return exception message as response.

**Review by Ehtesham599 — CHANGES_REQUESTED** [review]

**Comment by Amna-Alshawa** [comment]
@Ehtesham599 ,kindly review my PR.

File: one_fm/api/mobile/shift_permission.py [hunk 6]
```diff
@@ -31,17 +31,17 @@ def create_shift_permission(employee, permission_type, date, reason, leaving_tim
...
+                return response("{0} has already applied for permission to {1} on {2}.".format(doc.emp_name, permission_type.lower(), date), {}, False, 409)
 
         elif not shift or not type or not assigned_shift or not shift_supervisor:
-            return response({'message':"Shift Details are missing, Please Make sure You are have a Scheduled record on {0}".format(date),'data':{}},400)
+            return response("Could Not Create Shift Permission, Shift Details are missing.", {}, False, 400)
            
     except Exception as e:
         frappe.log_error(frappe.get_traceback())
-        return response({'message':"Shift Permission Not Created Successfully",'data':{}},500)
+        return frappe.utils.response.report_error(e)
```
Comment (Ehtesham599) [inline one_fm/api/mobile/shift_permission.py]: Use `return response(e, {}, False, 500)`

File: one_fm/api/mobile/shift_permission.py [hunk 7]
```diff
@@ -116,14 +116,14 @@ def shift_permission_approved(supervisor_id, shift_permission_id):
...
             elif shift_permission_record.workflow_state == "Approved":
-                return response({'message':"Shift Permission is Already Approved",'data':{}},400)
+                return response("Shift Permission is Already Approved", {}, False, 400)
         elif shift_supervisor != supervisor_id:
-            return response({'message':"Only Supervisor Has The Right to Approve",'data':{}},400)
+            return response("Only Supervisor Has The Right to Approve", {}, False, 400)
     except Exception as e:
         frappe.log_error(frappe.get_traceback())
-        return response({'message':"Shift Permission Not Approved Successfully",'data':{}},500)
+        return frappe.utils.response.report_error(e)
```
Comment (Ehtesham599) [inline one_fm/api/mobile/shift_permission.py]: Use `return response(e, {}, False, 500)`

File: one_fm/api/mobile/shift_permission.py [hunk 8]
```diff
@@ -142,14 +142,14 @@ def shift_permission_rejected(supervisor_id, shift_permission_id):
...
             elif shift_permission_record.workflow_state == "Rejected":
-                return response({'message':"Shift Permission is Already Rejected",'data':{}},400)
+                return response("Shift Permission is Already Rejected", {}, False, 400)
         elif shift_supervisor != supervisor_id:
-            return response({'message':"Only Supervisor Has The Right to Reject",'data':{}},400)
+            return response("Only Supervisor Has The Right to Reject", {}, False, 400)
     except Exception as e:
         frappe.log_error(frappe.get_traceback())
-        return response({'message':"Shift Permission Not Rejected Successfully",'data':{}},500)
+        return frappe.utils.response.report_error(e)
```
Comment (Ehtesham599) [inline one_fm/api/mobile/shift_permission.py]: Use `return response(e, {}, False, 500)`

File: one_fm/api/mobile/shift_permission.py [hunk 9]
```diff
@@ -169,14 +169,14 @@ def shift_permission_cancelled(supervisor_id, shift_permission_id):
...
             elif shift_permission_record.workflow_state == "Cancelled":
-                return response({'message':"Shift Permission is Already Cancelled",'data':{}},400)
+                return response("Shift Permission is Already Cancelled", {}, False, 400)
         elif shift_supervisor != supervisor_id:
-            return response({'message':"Only Supervisor Has The Right to Cancel",'data':{}},400)
+            return response("Only Supervisor Has The Right to Cancel", {}, False, 400)
     except Exception as e:
         frappe.log_error(frappe.get_traceback())
-        return response({'message':"Shift Permission Not Cancelled Successfully",'data':{}},500)
+        return frappe.utils.response.report_error(e)
```
Comment (Ehtesham599) [inline one_fm/api/mobile/shift_permission.py]: Use` return response(e, {}, False, 500)`

File: one_fm/api/notification.py [hunk 10]
```diff
@@ -19,22 +19,30 @@ def get_employee_user_id(employee):
...
-		else:
-
-			return ('No Notification.')
+			return response("Notifications Are Listed Sucessfully", notification_list, True, 200)
 
 	except Exception as e:
 		frappe.log_error(frappe.get_traceback())
-		return frappe.utils.response.report_error(e)
\ No newline at end of file
+		return frappe.utils.response.report_error(e)
```
Comment (Ehtesham599) [inline one_fm/api/notification.py]: Use `return response(e, {}, False, 500)`

File: one_fm/api/notification.py [hunk 11]
```diff
@@ -19,22 +19,30 @@ def get_employee_user_id(employee):
...
 	"""
 	try:
-		notification_list = frappe.get_all("Notification Log", filters={'for_user':user_id, 'one_fm_mobile_app':1}, fields=["name","subject"])
+		notification_list = frappe.get_all("Notification Log", filters={'for_user':frappe.session.user, 'one_fm_mobile_app':1}, fields=["name","subject"])
 		if len(notification_list)>0:
-			return notification_list 
-		else:
-
-			return ('No Notification.')
+			return response("Notifications Are Listed Sucessfully", notification_list, True, 200)
```
Comment (Ehtesham599) [inline one_fm/api/notification.py:36]: Please add appropriate response for the else condition (Where the length of notification list is 0).

File: one_fm/api/notification.py [hunk 12]
```diff
@@ -19,22 +19,30 @@ def get_employee_user_id(employee):
...
 	"""
 	Params:
-	User id: Employee User ID
-	Return: List of Notification
+	employee: employee ERP id 
+	Returns: notification list
 	"""
 	try:
-		notification_list = frappe.get_all("Notification Log", filters={'for_user':user_id, 'one_fm_mobile_app':1}, fields=["name","subject"])
+		notification_list = frappe.get_all("Notification Log", filters={'for_user':frappe.session.user, 'one_fm_mobile_app':1}, fields=["name","subject"])
```
Comment (Ehtesham599) [inline one_fm/api/notification.py]: Pass user_id itself. 
`frappe.session.user` will result in "Administrator"!.

**Review by Ehtesham599 — CHANGES_REQUESTED** [review]

**Review by Ehtesham599 — APPROVED** [review]
@Amna-onefm LGTM. Great work!

**Pull request merged**